flink-connectors/flink-connector-rabbitmq2/src/main/java/org/apache/flink/connector/rabbitmq2/sink/writer/specalized/RabbitMQSinkWriterAtLeastOnce.java

```diff
@@ -0,0 +1,192 @@
+            List<RabbitMQSinkWriterState<T>> states) {
+        super(
+                connectionConfig,
+                queueName,
+                serializationSchema,
+                publishOptions,
+                maxRetry,
+                returnListener);
+        this.outstandingConfirms = new ConcurrentSkipListMap<>();
+        this.lastSeenMessageIds = new HashSet<>();
+        this.lastResendTimestampMilliseconds = System.currentTimeMillis();
+        this.resendIntervalMilliseconds =
+                minimalResendIntervalMilliseconds != null
+                        ? minimalResendIntervalMilliseconds
+                        : DEFAULT_MINIMAL_RESEND_INTERVAL;
+        initWithState(states);
+    }
+
+    private void initWithState(List<RabbitMQSinkWriterState<T>> states) {
+        for (RabbitMQSinkWriterState<T> state : states) {
+            for (SinkMessage<T> message : state.getOutstandingMessages()) {
+                send(message);
+            }
+        }
+    }
+
+    @Override
+    protected void send(SinkMessage<T> msg) {
+        long sequenceNumber = rmqChannel.getNextPublishSeqNo();
+        super.send(msg);
+        outstandingConfirms.put(sequenceNumber, msg);
+    }
+
+    private void resendMessages() {
+        Set<Long> temp = outstandingConfirms.keySet();
+        Set<Long> messagesToResend = new HashSet<>(temp);
+        messagesToResend.retainAll(lastSeenMessageIds);
+        for (Long id : messagesToResend) {
+            // remove the old message from the map, since the message was added a second time
+            // under a new id or is put into the list of messages to resend
+            SinkMessage<T> msg = outstandingConfirms.remove(id);
+            if (msg != null) {
+                send(msg);
+            }
+        }
+        lastSeenMessageIds = temp;
+    }
+
+    private ConfirmCallback handleAcknowledgements() {
+        return (sequenceNumber, multiple) -> {
+            // multiple flag indicates that all messages < sequenceNumber can be safely acknowledged
+            if (multiple) {
+                // create a view of the portion of the map that contains keys < sequenceNumber
+                ConcurrentNavigableMap<Long, SinkMessage<T>> confirmed =
+                        outstandingConfirms.headMap(sequenceNumber, true);
+                // changes to the view are reflected in the original map
+                confirmed.clear();
+            } else {
+                outstandingConfirms.remove(sequenceNumber);
+            }
+        };
+    }
+
+    private ConfirmCallback handleNegativeAcknowledgements() {
+        return (sequenceNumber, multiple) -> {
+            SinkMessage<T> message = outstandingConfirms.get(sequenceNumber);
+            LOG.error(
+                    "Message with body {} has been nack-ed. Sequence number: {}, multiple: {}",
+                    message.getMessage(),
+                    sequenceNumber,
+                    multiple);
+        };
```

Comment:
From RabbitMQ docs: 

_"Reject one or more incoming messages.
This method allows a client to reject one or more incoming messages. It can be used to interrupt and cancel large incoming messages, or return untreatable messages to their original queue. This method is also used by the server to inform publishers on channels in confirm mode of unhandled messages. If a publisher receives this method, it probably needs to republish the offending messages."_

The basic.nack method indicates the message somehow cannot be processed by the RabbitMQ server. 
IMO, for at-most-once it should be okay to keep the job running if individual messages do not arrive, but for at-least and exactly-once it should probably fail.

Reply:
Done